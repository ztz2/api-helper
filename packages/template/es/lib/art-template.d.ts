import artTemplate from './utils/template/art-template-web.js';
export default artTemplate;
