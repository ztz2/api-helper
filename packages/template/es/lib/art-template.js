var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
// @ts-ignore
import artTemplate from '../lib/utils/template/art-template-web.js';
// @ts-ignore
artTemplate.defaults.escape = false;
artTemplate.defaults.minimize = false;
// 新增模板《》语法的界定符规则
artTemplate.defaults.rules[2] = __assign(__assign({}, artTemplate.defaults.rules[1]), { test: /《([@#]?)[ \t]*(\/?)([\w\W]*?)[ \t]*》/ });
artTemplate.defaults.imports.renderCharacterByBoolean = function (booleanStatus, trueValue, falseValue) {
    if (booleanStatus === void 0) { booleanStatus = true; }
    if (trueValue === void 0) { trueValue = ''; }
    if (falseValue === void 0) { falseValue = ''; }
    return booleanStatus ? trueValue : falseValue;
};
artTemplate.defaults.imports.renderTemplateLiteral = function (booleanStatus, defaultValue) {
    if (booleanStatus === void 0) { booleanStatus = true; }
    if (defaultValue === void 0) { defaultValue = '\''; }
    return booleanStatus ? '`' : defaultValue;
};
export default artTemplate;
