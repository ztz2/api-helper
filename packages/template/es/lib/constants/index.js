export var API_CUSTOM_TEMPLATE_ID = 'API自定义分组ID_ea18930468d9402e';
export var MODEL_CUSTOM_TEMPLATE_ID = '模型自定义分组ID_6f3f32d6b404aaf9';
export var DEFAULT_SELECT_API_TPL_ID = 'typescript_api_76ea205dba9f3dfb';
export var DEFAULT_SELECT_MODEL_TPL_ID = 'javascript_map_841d8ffc9b34d1d5';
