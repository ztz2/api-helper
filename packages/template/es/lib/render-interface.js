import merge from 'lodash/merge';
import cloneDeep from 'lodash/cloneDeep';
import * as _changeCase from 'change-case';
import { createSchema } from '@api-helper/core/lib/helpers';
import { randomChar, processKeyName } from '@api-helper/core/lib/utils/util';
import { postCode, isEmptyObject, checkIsInterface, } from '../lib/utils/util';
import artTemplate from '../lib/art-template';
import { precessArraySchema } from '../lib/render-object';
export function renderInterface(schema, api, options) {
    options = merge({
        onlyBody: false,
        prefix: 'export ',
        paramType: 'request',
        emptyBodyCode: 'any;',
    }, options);
    schema = cloneDeep(schema);
    schema = precessArraySchema(schema);
    var sourceSchema = schema;
    var prefix = options.prefix, onlyBody = options.onlyBody, dropComment = options.dropComment, isExtraData = options.isExtraData, emptyBodyCode = options.emptyBodyCode, _a = options.paramType, paramType = _a === void 0 ? 'request' : _a;
    if (Array.isArray(schema)) {
        schema = createSchema('object', {
            id: randomChar(),
            params: schema,
        });
    }
    var isInterface = checkIsInterface(schema);
    var keyword = isInterface ? "".concat(prefix, " interface") : "".concat(prefix, "type");
    var onRenderInterfaceName = (options === null || options === void 0 ? void 0 : options.onRenderInterfaceName) ? options.onRenderInterfaceName : renderInterfaceName;
    var commentCode = onlyBody ? '' : dropComment !== true ? renderInterfaceComment(api, paramType, isExtraData) : '';
    var interfaceName = (options === null || options === void 0 ? void 0 : options.name) ? options.name : onRenderInterfaceName(api, {
        schema: schema,
        paramType: paramType,
        isExtraData: isExtraData,
        changeCase: _changeCase,
    });
    /**
     * output ->  export interface interfaceName
     *            export type Type =
     */
    var ki = ["".concat(keyword, " ").concat(interfaceName, " ").concat(!isInterface ? '=' : '')].filter(Boolean).join('\n');
    // 前置处理，渲染数据为空，直接返回 export type xxx = any;
    if ((Array.isArray(sourceSchema) && sourceSchema.length === 0) ||
        !sourceSchema ||
        isEmptyObject(sourceSchema)) {
        // 不兼容的数据，不显示类型
        if (isExtraData) {
            return '';
        }
        return postCode({
            ki: ki,
            commentCode: commentCode,
            code: emptyBodyCode
        }, { onlyBody: onlyBody });
    }
    if (paramType === 'response') {
        schema.keyName = '';
    }
    return postCode({
        ki: ki,
        commentCode: commentCode,
        code: renderInterfaceDeepObject(schema, true)
    }, { onlyBody: onlyBody });
}
export function renderInterfaceName(api, options) {
    var name = api.path;
    if (options.paramType) {
        name += " ".concat(options.paramType);
    }
    if (options.isExtraData) {
        name += 'ExtraData';
    }
    name += "By ".concat(api.method);
    return options.changeCase.pascalCase(name);
}
function renderInterfaceComment(api, paramType, isExtraData, updateTime) {
    if (isExtraData === void 0) { isExtraData = false; }
    if (updateTime === void 0) { updateTime = ''; }
    return artTemplate.render("/**\n * @description \u300Adescription\u300B\u300Aif docURL\u300B\n * @doc \u300AdocURL\u300B\u300A/if\u300B\n * @url \u300Aurl\u300B\u300Aif updateTime\u300B\n * @update \u300AupdateTime\u300B\u300A/if\u300B\n */", {
        description: "".concat([api.title, api.description].filter(Boolean).join('、'), "\u3010").concat(isExtraData ? '不兼容的请求数据' : paramType === 'request' ? '请求数据' : paramType === 'response' ? '响应数据' : '', "\u7C7B\u578B\u5B9A\u4E49\u3011"),
        docURL: api.docURL,
        url: "[ ".concat(api.method.toUpperCase(), " ] ").concat(api.path),
        updateTime: updateTime
    });
}
function requiredChar(schema) {
    var _a;
    return !((_a = schema === null || schema === void 0 ? void 0 : schema.rules) === null || _a === void 0 ? void 0 : _a.required) ? '?' : '';
}
function renderInterfaceDeepObject(schema, isRoot, memo) {
    if (isRoot === void 0) { isRoot = false; }
    if (memo === void 0) { memo = new Map(); }
    if (!schema) {
        return '';
    }
    if (memo.has(schema)) {
        return memo.get(schema);
    }
    memo.set(schema, 'null');
    var type = schema.type;
    var bannerComment = [];
    if (schema.label) {
        bannerComment.push(schema.label);
    }
    var bannerCommentText = bannerComment.length > 0 ? "\n// ".concat(bannerComment.join('')) : '';
    if (isEmptyObject(schema)) {
        return [
            bannerCommentText,
            schema.keyName ? "".concat(processKeyName(schema.keyName)).concat(requiredChar(schema), ": ") : '',
            '{',
            '[propName: string]: any',
            '}',
        ].filter(Boolean).join('\n');
    }
    var code = '';
    switch (type) {
        // 对象类型
        case 'object':
            code = [
                bannerCommentText,
                schema.keyName ? "".concat(processKeyName(schema.keyName)).concat(requiredChar(schema), ": ") : '',
                '{',
                // 类型遍历
                schema.params.filter(function (item) { var _a; return !(((_a = item.keyName) === null || _a === void 0 ? void 0 : _a.trim()) === '' && item.type === 'object'); }).map(function (item) { return renderInterfaceDeepObject(item, false, memo); }).join('\n'),
                '}',
            ].filter(Boolean).join('\n');
            break;
        // 数据类型
        case 'array':
            var child = schema.params.map(function (item) { return renderInterfaceDeepObject(item, false, memo); }).join(' | ');
            child = child || 'any';
            code = [
                bannerCommentText,
                schema.keyName ? "".concat(processKeyName(schema.keyName)).concat(requiredChar(schema), ": ") : '',
                // 类型遍历
                "Array<".concat(child, ">"),
            ].filter(Boolean).join('\n');
            break;
        // 其他非引用类型
        default:
            // eslint-disable-next-line no-case-declarations
            var typeCode = type;
            if (type === 'string' && 'enum' in schema && schema.enum.length) {
                typeCode = schema.enum.map(function (item) { return "'".concat(item, "'"); }).join(' | ');
            }
            if (schema.keyName) {
                code = [bannerCommentText, "".concat(processKeyName(schema.keyName)).concat(requiredChar(schema), ": ").concat(typeCode)].filter(Boolean).join('\n');
            }
            else {
                code = typeCode;
            }
    }
    if (isRoot && schema.type !== 'object' && schema.keyName) {
        code = "{ \n ".concat(code, " \n } \n");
    }
    return code;
}
