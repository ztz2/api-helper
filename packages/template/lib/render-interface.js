"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkInterface = exports.renderInterfaceName = exports.renderInterface = void 0;
var changeCase = __importStar(require("change-case"));
var prettier_1 = __importDefault(require("../lib/utils/prettier"));
function renderInterface(schema, api, options) {
    var isExtraData = options.isExtraData, paramType = options.paramType, hasComment = options.hasComment;
    var keyword = checkInterface(schema) ? 'export interface ' : 'export type ';
    var onRenderInterfaceName = (options === null || options === void 0 ? void 0 : options.onRenderInterfaceName) ? options.onRenderInterfaceName : renderInterfaceName;
    var commentCode = hasComment !== false ? renderInterfaceComment(schema, api, paramType, isExtraData) : '';
    var interfaceName = onRenderInterfaceName(schema, api, {
        isExtraData: isExtraData,
        paramType: paramType,
        changeCase: changeCase
    });
    // 前置处理
    if (!schema || isEmptyObject(schema)) {
        // 不兼容的数据，不显示类型
        if (isExtraData) {
            return '';
        }
        return [
            commentCode,
            "".concat(keyword, " ").concat(interfaceName, " = any;")
        ].filter(Boolean).join('\n');
    }
    if (paramType === 'response') {
        schema.keyName = '';
    }
    var code = renderInterfaceDeepObject(schema);
    // 移除多余换行符
    code = code.replace(/\n\n/gim, '\n');
    // 开始组合
    interfaceName = keyword + interfaceName;
    if (schema.type !== 'object') {
        interfaceName += ' = ';
    }
    var formatted = (0, prettier_1.default)(interfaceName + code, {
        parser: 'typescript'
    });
    return [
        commentCode,
        formatted
    ].filter(Boolean).join('\n');
}
exports.renderInterface = renderInterface;
function renderInterfaceName(schema, api, options) {
    var isInterface = checkInterface(schema);
    var name = isInterface ? 'I' : '';
    name += api.path;
    if (options.paramType) {
        name += " ".concat(options.paramType);
    }
    if (!isInterface) {
        name += 'Type';
    }
    if (options.isExtraData) {
        name += 'ExtraData';
    }
    name += "By ".concat(api.method);
    return changeCase.pascalCase(name);
}
exports.renderInterfaceName = renderInterfaceName;
function renderInterfaceComment(schema, api, paramType, isExtraData) {
    if (isExtraData === void 0) { isExtraData = false; }
    var commentText = "/**\n * @description ".concat([api.title, api.description].filter(Boolean).join('、'), "\u3010").concat(isExtraData ? '不兼容的请求数据' : paramType === 'request' ? '请求数据' : paramType === 'response' ? '响应数据' : '', "\u7C7B\u578B\u5B9A\u4E49\u3011").concat(api.docURL ? "\n * @doc ".concat(api.docURL, "\n") : '', "\n * @url [ ").concat(api.method.toUpperCase(), " ] ").concat(api.path, "\n */");
    return commentText;
}
function renderInterfaceDeepObject(schema, memo) {
    if (memo === void 0) { memo = new Map(); }
    if (!schema) {
        return '';
    }
    if (memo.has(schema)) {
        return memo.get(schema);
    }
    memo.set(schema, 'null');
    var type = schema.type;
    var bannerComment = [];
    if (schema.label) {
        bannerComment.push(schema.label);
    }
    var bannerCommentText = bannerComment.length > 0 ? "\n// ".concat(bannerComment.join('')) : '';
    if (isEmptyObject(schema)) {
        return [
            bannerCommentText,
            schema.keyName ? "".concat(schema.keyName, ": ") : '',
            '{',
            '[propName: string]: any',
            '}',
        ].filter(Boolean).join('\n');
    }
    var code = '';
    switch (type) {
        // 对象类型
        case 'object':
            code = [
                bannerCommentText,
                schema.keyName ? "".concat(schema.keyName, ": ") : '',
                '{',
                // 类型遍历
                schema.params.map(function (item) { return renderInterfaceDeepObject(item, memo); }).join('\n'),
                '}',
            ].filter(Boolean).join('\n');
            break;
        // 数据类型
        case 'array':
            var child = schema.params.map(function (item) { return renderInterfaceDeepObject(item, memo); }).join(' | ');
            child = child || 'any';
            code = [
                bannerCommentText,
                schema.keyName ? "".concat(schema.keyName, ": ") : '',
                // 类型遍历
                "Array<".concat(child, ">"),
            ].filter(Boolean).join('\n');
            break;
        // 其他非引用类型
        default:
            // eslint-disable-next-line no-case-declarations
            var typeCode = type;
            if (type === 'string' && 'enum' in schema && schema.enum.length) {
                typeCode = schema.enum.map(function (item) { return "'".concat(item, "'"); }).join(' | ');
            }
            if (schema.keyName) {
                code = [bannerCommentText, "".concat(schema.keyName).concat(!schema.rules.required ? '?' : '', ": ").concat(typeCode)].filter(Boolean).join('\n');
            }
            else {
                code = typeCode;
            }
    }
    return code;
}
function checkInterface(schema) {
    var _a;
    var isType = !schema || (schema === null || schema === void 0 ? void 0 : schema.type) === 'array' || isEmptyObject(schema);
    var hasKey = (_a = schema === null || schema === void 0 ? void 0 : schema.params) === null || _a === void 0 ? void 0 : _a.some(function (s) { return s.keyName; });
    return !isType && hasKey;
}
exports.checkInterface = checkInterface;
function isEmptyObject(schema) {
    var _a;
    return (schema === null || schema === void 0 ? void 0 : schema.type) === 'object' && ((_a = schema === null || schema === void 0 ? void 0 : schema.params) === null || _a === void 0 ? void 0 : _a.length) === 0;
}
