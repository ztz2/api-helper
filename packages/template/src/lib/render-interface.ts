import merge from 'lodash/merge';
import cloneDeep from 'lodash/cloneDeep';
import * as _changeCase from 'change-case';
import { ChangeCase } from '@/lib/types';
import { APIHelper } from '@api-helper/core/lib/types';
import { uuid } from '@api-helper/core/lib/utils/util';
import { createSchema } from '@api-helper/core/lib/helpers';

import { checkIsInterface, isEmptyObject, postCode } from '@/lib/utils/util';

export function renderInterface(
  schema: APIHelper.Schema | Array<APIHelper.Schema> | null,
  api: APIHelper.API,
  options?: {
    // 请求数据还是响应数据
    paramType?: 'request' | 'response';
    // 前缀
    prefix?: string,
    // interface 名称
    name?: string;
    // 删除interface注释
    dropComment?: boolean;
    // 只生成类型部分
    onlyBody?: boolean;
    // 生成接口名称回调
    onRenderInterfaceName?: typeof renderInterfaceName;
    // 是否是额外数据
    isExtraData?: boolean;
    // 当数据为空的body代码
    emptyBodyCode?: string;
  }) {
  options = merge({
    onlyBody: false,
    prefix: 'export ',
    paramType: 'request',
    emptyBodyCode: 'any;',
  }, options);

  schema = cloneDeep(schema);
  const sourceSchema = schema;

  const {
    prefix,
    onlyBody,
    dropComment,
    isExtraData,
    emptyBodyCode,
    paramType = 'request',
  } = options;

  if (Array.isArray(schema)) {
    schema = createSchema('object', {
      id: uuid(),
      params: schema,
    }) as APIHelper.Schema;
  }

  const isInterface = checkIsInterface(schema);
  const keyword = isInterface ? `${prefix} interface` : `${prefix}type`;
  const onRenderInterfaceName = options?.onRenderInterfaceName ? options.onRenderInterfaceName : renderInterfaceName;
  let commentCode = onlyBody ? '' : dropComment !== true ? renderInterfaceComment(schema, api, paramType, isExtraData) : '';
  let interfaceName = options?.name ? options.name : onRenderInterfaceName(schema, api, {
    isExtraData,
    paramType,
    changeCase: _changeCase,
  });

  /**
   * output ->  export interface interfaceName
   *            export type Type =
   */
  let ki = [`${keyword} ${interfaceName} ${!isInterface ? '=' : ''}`].filter(Boolean).join('\n');

  // 前置处理，渲染数据为空，直接返回 export type xxx = any;
  if ((Array.isArray(sourceSchema) && sourceSchema.length === 0) ||
    !sourceSchema ||
    isEmptyObject(sourceSchema as APIHelper.Schema)) {
    // 不兼容的数据，不显示类型
    if (isExtraData) {
      return '';
    }
    return postCode({
      ki,
      commentCode,
      code: emptyBodyCode
    }, { onlyBody });
  }

  if (paramType === 'response') {
    (schema as APIHelper.Schema).keyName = '';
  }

  return postCode({
    ki,
    commentCode,
    code: renderInterfaceDeepObject(schema as APIHelper.Schema)
  }, { onlyBody });
}

export function renderInterfaceName (
  schema: APIHelper.Schema | null,
  api: APIHelper.API,
  options: {
    isExtraData?: boolean;
    paramType: 'request' | 'response';
    changeCase: ChangeCase;
  }): string {
  const changeCase = options.changeCase ?? _changeCase;
  const isInterface = checkIsInterface(schema);

  let name = `${isInterface ? 'I' : ''}${api.path}`;

  if (options.paramType) {
    name += ` ${options.paramType}`;
  }
  if (!isInterface) {
    name += 'Type';
  }
  if (options.isExtraData) {
    name += 'ExtraData'
  }
  name += `By ${api.method}`;
  return changeCase.pascalCase(name);
}

function renderInterfaceComment(
  schema: APIHelper.Schema | null,
  api: APIHelper.API,
  paramType: 'request' | 'response',
  isExtraData = false
) {
  const commentText =
`/**
 * @description ${[api.title, api.description].filter(Boolean).join('、')}【${isExtraData ? '不兼容的请求数据' : paramType === 'request' ? '请求数据' : paramType === 'response' ? '响应数据' : ''}类型定义】${api.docURL ? `\n * @doc ${api.docURL}` : ''}
 * @url [ ${api.method.toUpperCase()} ] ${api.path}
 */`;
  return commentText;
}

function requiredChar(schema: APIHelper.Schema) {
  return !schema?.rules?.required ? '?' : '';
}

function renderInterfaceDeepObject(
  schema: APIHelper.Schema,
  memo = new Map<APIHelper.Schema, string>()
): string {
  if (!schema) {
    return '';
  }
  if (memo.has(schema)) {
    return memo.get(schema) as string;
  }

  memo.set(schema, 'null');

  const { type } = schema;
  const bannerComment = [];
  if (schema.label) {
    bannerComment.push(schema.label);
  }
  const bannerCommentText = bannerComment.length > 0 ? `\n// ${  bannerComment.join('')}` : '';

  if (isEmptyObject(schema)) {
    return [
      bannerCommentText,
      schema.keyName ? `${schema.keyName}${requiredChar(schema)}: ` : '',
      '{',
        '[propName: string]: any',
      '}',
    ].filter(Boolean).join('\n');
  }

  let code = '';
  switch (type) {
    // 对象类型
    case 'object':
      code = [
        bannerCommentText,
        schema.keyName ? `${schema.keyName}${requiredChar(schema)}: ` : '',
        '{',
          // 类型遍历
          schema.params.map((item) => renderInterfaceDeepObject(item, memo)).join('\n'),
        '}',
      ].filter(Boolean).join('\n');
      break;
    // 数据类型
    case 'array':
      let child = schema.params.map((item) => renderInterfaceDeepObject(item, memo)).join(' | ');
      child = child || 'any';
      code = [
        bannerCommentText,
        schema.keyName ? `${schema.keyName}${requiredChar(schema)}: ` : '',
        // 类型遍历
        `Array<${child}>`,
      ].filter(Boolean).join('\n');
      break;
    // 其他非引用类型
    default:
      // eslint-disable-next-line no-case-declarations
      let typeCode: string = type;
      if (type === 'string' && 'enum' in schema && schema.enum.length) {
        typeCode = schema.enum.map((item) => `'${item}'`).join(' | ');
      }
      if (schema.keyName) {
        code = [bannerCommentText, `${schema.keyName}${requiredChar(schema)}: ${typeCode}`].filter(Boolean).join('\n');
      } else {
        code = typeCode;
      }
  }

  return code;
}
