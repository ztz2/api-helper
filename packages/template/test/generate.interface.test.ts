import { join } from 'path';
import ParserSwagger from '@api-helper/core/es/lib/parser/parser-swagger';

import { readJsonFile } from './utils/file';
import { renderInterface } from '../src/lib/render-interface';

describe('生成 interface 测试', () => {
  test('open-api-响应参数为List', async () => {
    const openAPIDocumentJSON = await readJsonFile(join(__dirname, '../../core/test/resources/open-api-response-array.json'));
    const openAPIDocumentList = await new ParserSwagger(false).parser([openAPIDocumentJSON]);
    const openAPIDocument = openAPIDocumentList[0];
    const category = openAPIDocument.categoryList[0];
    const api = category.apiList[0];
    const responseCode = [
      renderInterface(api.requestDataSchema, api),
      renderInterface(api.responseDataSchema, api),
    ].join('\n');
    expect(
      responseCode,
    ).toMatchSnapshot('open-api-响应参数为List');
  });

  test('open-api-2.0-dot.json', async () => {
    const openAPIDocumentJSON = await readJsonFile(join(__dirname, '../../core/test/resources/open-api-2.0-dot.json'));
    const openAPIDocumentList = await new ParserSwagger(false).parser([openAPIDocumentJSON]);
    const openAPIDocument = openAPIDocumentList[0];
    const category = openAPIDocument.categoryList[0];
    const api = category.apiList[0];
    const responseCode = [
      renderInterface(api.requestDataSchema, api),
      renderInterface(api.responseDataSchema, api),
    ].join('\n');
    expect(
      responseCode,
    ).toMatchSnapshot('open-api-2.0-dot.json');
  });

  test('OpenAPI-2.0-source生成 interface 测试', async () => {
    const openAPIDocumentJSON = await readJsonFile(join(__dirname, '../../core/test/resources/open-api-2.0-source.json'));
    const openAPIDocumentList = await new ParserSwagger(false).parser([openAPIDocumentJSON]);
    const openAPIDocument = openAPIDocumentList[0];
    const category = openAPIDocument.categoryList[0];
    const api = category.apiList[0];
    const responseCode = renderInterface(api.responseDataSchema, api, {
      paramType: 'request',
      name: '安逗和黑子',
    });
    expect(
      responseCode,
    ).toMatchSnapshot('OpenAPI-2.0-source生成 interface 测试');
  });

  test('OpenAPI-2.0生成 interface 测试', async () => {
    const openAPIDocumentJSON = await readJsonFile(join(__dirname, '../../core/test/resources/open-api-2.0.json'));
    const openAPIDocumentList = await new ParserSwagger(false).parser([openAPIDocumentJSON]);
    const openAPIDocument = openAPIDocumentList[0];
    const category = openAPIDocument.categoryList[1];
    const api = category.apiList[0];
    const responseCode = renderInterface(api.responseDataSchema, api, {
      paramType: 'request',
      name: '安逗和黑子',
    });
    expect(
      responseCode,
    ).toMatchSnapshot('OpenAPI-2.0生成 interface 测试');
  });

  test('OpenAPI-3.0生成 interface 测试', async () => {
    const openAPIDocumentJSON = await readJsonFile(join(__dirname, '../../core/test/resources/open-api-3.0.json'));
    const openAPIDocumentList = await new ParserSwagger(false).parser([openAPIDocumentJSON]);
    const openAPIDocument = openAPIDocumentList[0];
    const category = openAPIDocument.categoryList[1];
    const api = category.apiList[1];
    const responseCode = renderInterface(api.requestDataSchema, api, {
      paramType: 'response',
    });
    expect(
      responseCode,
    ).toMatchSnapshot('OpenAPI-3.0生成 interface 测试');
  });

  test('name - prefix - dropComment - onlyBody 属性测试', async () => {
    const openAPIDocumentJSON = await readJsonFile(join(__dirname, '../../core/test/resources/open-api-2.0.json'));
    const openAPIDocumentList = await new ParserSwagger(false).parser([openAPIDocumentJSON]);
    const openAPIDocument = openAPIDocumentList[0];
    const category = openAPIDocument.categoryList[1];
    const api = category.apiList[0];
    expect(
      renderInterface(api.responseDataSchema, api, {
        paramType: 'request',
        name: '自定义名称',
      }),
    ).toMatchSnapshot('OpenAPI-2.0生成 interface - name属性测试');
    expect(
      renderInterface(api.responseDataSchema, api, {
        paramType: 'request',
        prefix: '',
      }),
    ).toMatchSnapshot('OpenAPI-2.0生成 interface - prefix属性测试');
    expect(
      renderInterface(api.responseDataSchema, api, {
        paramType: 'request',
        dropComment: true,
      }),
    ).toMatchSnapshot('OpenAPI-2.0生成 interface - dropComment属性测试');
    expect(
      renderInterface([{ "id":"db90e419-44ff-4ea9-95e1-c68f00b78273","type":"string","keyName":"phone","title":"","description":"手机号","label":"手机号","rules":{ "required":true },"examples":[],"params":[],"enum":[] },{ "id":"919d1555-82af-4b01-b4b4-4e16062a0a42","type":"string","keyName":"username","title":"","description":"账户名","label":"账户名","rules":{ "required":true },"examples":[],"params":[],"enum":[] }], api, {
        prefix: '',
        name: 'RuleForm',
        onlyBody: true,
      }),
    ).toMatchSnapshot('OpenAPI-2.0生成 interface - onlyBody属性测试');
  });
});
