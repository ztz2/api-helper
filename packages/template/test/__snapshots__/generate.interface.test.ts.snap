// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`生成 interface 测试 OpenAPI-2.0生成 interface 测试: OpenAPI-2.0生成 interface 测试 1`] = `
"/**
 * @description uploads an image
 * @summary Request data types
 * @url [ POST ] /v2/pet/{petId}/uploadImage
 */
export interface 安逗和黑仔 {
code: number
type: string
message: string
}"
`;

exports[`生成 interface 测试 OpenAPI-3.0生成 interface 测试: OpenAPI-3.0生成 interface 测试 1`] = `
"/**
 * @description Place an order for a pet、Place a new order in the store
 * @summary Response data types
 * @url [ POST ] /api/v3/store/order
 */
export interface ApiV3StoreOrderResponseByPost {
id?: string
petId?: string
quantity?: number
shipDate?: string
// Order Status
status?: 'placed' | 'approved' | 'delivered'
complete?: boolean
}"
`;

exports[`生成 interface 测试 name - prefix - dropComment - onlyBody 属性测试: OpenAPI-2.0生成 interface - dropComment属性测试 1`] = `"export type V2StoreInventoryRequestByGet =any;"`;

exports[`生成 interface 测试 name - prefix - dropComment - onlyBody 属性测试: OpenAPI-2.0生成 interface - name属性测试 1`] = `
"/**
 * @description Returns pet inventories by status、Returns a map of status codes to quantities
 * @summary Request data types
 * @url [ GET ] /v2/store/inventory
 */
export type 自定义名称 =any;"
`;

exports[`生成 interface 测试 name - prefix - dropComment - onlyBody 属性测试: OpenAPI-2.0生成 interface - onlyBody属性测试 1`] = `
"{
// 手机号
phone: string
// 账户名
username: string
}"
`;

exports[`生成 interface 测试 name - prefix - dropComment - onlyBody 属性测试: OpenAPI-2.0生成 interface - prefix属性测试 1`] = `
"/**
 * @description Returns pet inventories by status、Returns a map of status codes to quantities
 * @summary Request data types
 * @url [ GET ] /v2/store/inventory
 */
type V2StoreInventoryRequestByGet =any;"
`;

exports[`生成 interface 测试 open-api-3.0-dot.json: open-api-3.0-dot.json 1`] = `
"/**
 * @description 兼容非标准的dot参数形式
 * @summary Request data types
 * @url [ GET ] /compatible-error-specification-dot-query-param
 */
export interface CompatibleErrorSpecificationDotQueryParamRequestByGet {
// ID集合
ids?: 
Array<string>
dept?: 
{
// 创建者1号
deptCreateBy1?: string
// 创建者2号
deptCreateBy2?: string
}
roles?: 
Array<{
// 角色创建者1号
roleCreateBy1?: string
// 角色创建时间
roleCreateTime?: string
}>
// 有分页参数的话就加,根据你的实际查询参数编写
"有分页参数的话就加,根据你的实际查询参数编写"?: string
}
/**
 * @description 兼容非标准的dot参数形式
 * @summary Request data types
 * @url [ GET ] /compatible-error-specification-dot-query-param
 */
export type CompatibleErrorSpecificationDotQueryParamRequestByGet =any;"
`;

exports[`生成 interface 测试 open-api-响应参数为List: open-api-响应参数为List 1`] = `
"/**
 * @description 响应数据 array
 * @summary Request data types
 * @url [ POST ] /resp-array
 */
export type RespArrayRequestByPost =any;
/**
 * @description 响应数据 array
 * @summary Request data types
 * @url [ POST ] /resp-array
 */
export type RespArrayRequestByPost =Array<{
// 用户名
username: string
}>"
`;
